netty/src/main/java/io/grpc/netty/NettyServerHandler.java

```diff
@@ -188,16 +188,17 @@ private void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers
       // method.
       Http2Stream http2Stream = requireHttp2Stream(streamId);
 
-      NettyServerStream stream = new NettyServerStream(ctx.channel(), http2Stream, this,
-              maxMessageSize);
+      NettyServerStream.TransportState state =
+          new NettyServerStream.TransportState(this, http2Stream, maxMessageSize);
+      NettyServerStream stream = new NettyServerStream(ctx.channel(), state);
 
       Metadata metadata = Utils.convertHeaders(headers);
-      stream.inboundHeadersReceived(metadata);
 
       ServerStreamListener listener =
           transportListener.streamCreated(stream, method, metadata);
-      stream.setListener(listener);
-      http2Stream.setProperty(streamKey, stream);
+      // TODO(ejona): this is racy since stream could have been used before getting here.
+      state.setListener(listener);
```

Comment:
Thinking about it a bit more. I don't think this is racy. The only way the stream may be used before the listener was set, is that `stream.close()` is called. However, this will just add a command to the write queue, which is drained by the same eventloop thread as the listener is set on. So that seems safe.